### PR TITLE
[WIP] Add cetl::variant

### PR DESCRIPTION
**Pull request by pavel-kirienko**
This is just an ordinary variant, nothing special here.

Other changes:

- In CETLVaSt, move special member function policies from the optional test into a separate header to make them reusable for the variant test.
- Extend the typelist tools in CETLVaSt.
- Add standard type traits from C++17 (I needed them for the variant): `conjunction_v` etc.
- Add the `overloaded` helper class, which is not found in the standard library but is commonly used with visitation. We can move it somewhere if its current place is found unsuitable.
- Add `type_traits_ext.hpp` with type traits intended for internal use mostly.
- Change the SonarQube invocation to run it against an exception-enabled build to provide full coverage information.
- Doxygen:
  - Ignore the `detail` namespace.
  - Disable `WARN_AS_ERROR` because Doxygen v1.9 is unable to parse certain recursive templates correctly; see https://stackoverflow.com/a/77855566/1007777

### CODE REVIEW

File: include/cetl/type_traits_ext.hpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+template <template <typename...> class Predicate, std::size_t Ix, typename Head, typename... Tail>
+struct find_impl<Predicate, Ix, Head, Tail...> : find_impl<Predicate, Ix + 1, Tail...>
+{
+    using base                                     = find_impl<Predicate, Ix + 1, Tail...>;
+    static constexpr std::size_t first_match_index = (Predicate<Head>::value ? Ix : base::first_match_index);
+    static constexpr std::size_t match_count       = (Predicate<Head>::value ? 1 : 0) + base::match_count;
+};
+}  // namespace detail
+
+/// The \c value member contains the index of the first type T in the typelist for which Predicate<T>::value is true.
```
Comment (thirtytwobits) [inline include/cetl/type_traits_ext.hpp:42]: Examples for find and count would be helpful.

**Review by thirtytwobits — DISMISSED** [review]

File: include/cetl/pf17/variant.hpp [hunk 2]
```diff
@@ -0,0 +1,1270 @@
...
+/// Implementation of \ref std::variant_size_v.
+/// This implementation also accepts other typelist-parameterized classes, such as \ref std::variant.
+template <typename V>
+constexpr size_t variant_size_v = variant_size<V>::value;
+
+// --------------------------------------------------------------------------------------------------------------------
+
+/// An implementation of \ref std::variant.
+///
+/// The valueless state cannot occur unless exceptions are enabled.
```
Comment (thirtytwobits) [inline include/cetl/pf17/variant.hpp:849]: Missing docs for @param T

File: include/cetl/pf17/variant.hpp [hunk 3]
```diff
@@ -0,0 +1,1270 @@
+/// @file
```
Comment (thirtytwobits) [inline include/cetl/pf17/variant.hpp:1]: Don't forget to hook this up to the polyfill headers.

File: include/cetl/pf17/variant.hpp [hunk 4]
```diff
@@ -0,0 +1,1270 @@
...
+// --------------------------------------------------------------------------------------------------------------------
+
+/// An implementation of \ref std::variant.
+///
+/// The valueless state cannot occur unless exceptions are enabled.
+///
+/// In this implementation, the address of the variant object is equivalent to the address of the active alternative;
+/// this can sometimes be useful for low-level debugging.
+template <typename... Ts>
+class variant : private detail::var::base_move_assignment<detail::var::types<Ts...>>
```
Comment (thirtytwobits) [inline include/cetl/pf17/variant.hpp:854]: This doesn't work as expected:
```
std::variant<std::string, void const*> y("abc");

error: no matching constructor for initialization of 'cetl::pf17::variant<std::string, const void *>' (aka 'variant<basic_string<char>, const void *>')
```

Comment (pavel-kirienko) [reply include/cetl/pf17/variant.hpp:854]: Yes, this is why this is still a draft. I found a divergence in the behavior of the converting ctor yesterday and am working on it now.

File: include/cetl/pf17/variant.hpp [hunk 5]
```diff
@@ -0,0 +1,1270 @@
...
+/// Invokes the specified fun as follows and forwards its result to the caller (argument forwarding not shown):
+/// @code
+/// fun(std::integral_constant<std::size_t, index>{}, extra_args...)
+/// @endcode
+/// In other words, this function moves the `index` value from the run time to the compile time domain (hence the name).
+/// `N` specifies the maximum value for the index, which also informs the size of the lookup table.
+/// If `index>=N`, behaves like `std::array<...,N>::at(index)`.
+/// The time complexity is constant.
+template <std::size_t N, typename F, typename... Args>
+constexpr decltype(auto) chronomorphize(F&& fun, const std::size_t index, Args&&... extra_args)
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply include/cetl/pf17/variant.hpp:107]: it is now

File: include/cetl/pf17/variant.hpp [hunk 6]
```diff
@@ -0,0 +1,1270 @@
...
+/// Implementation of \ref std::monostate.
+struct monostate
+{};
+// clang-format off
+inline constexpr bool operator==(const monostate, const monostate) noexcept { return true;  }
+inline constexpr bool operator!=(const monostate, const monostate) noexcept { return false; }
+inline constexpr bool operator< (const monostate, const monostate) noexcept { return false; }
+inline constexpr bool operator> (const monostate, const monostate) noexcept { return false; }
+inline constexpr bool operator<=(const monostate, const monostate) noexcept { return true;  }
+inline constexpr bool operator>=(const monostate, const monostate) noexcept { return true;  }
```
[the remark this answers is not in the document]
Comment (pavel-kirienko) [reply include/cetl/pf17/variant.hpp:41]: Moving this into a separate issue: https://github.com/OpenCyphal/CETL/issues/93

I omitted hash specializations on the assumption that an embedded application that needs CETL is unlikely to make use of probabilistic containers.

File: include/cetl/pf17/variant.hpp (hunk 4 above)
Comment (pavel-kirienko) [reply include/cetl/pf17/variant.hpp:829]: Fixing this was quite a trip.

File: include/cetl/pf17/variant.hpp (hunk 2 above)
Comment (pavel-kirienko) [reply include/cetl/pf17/variant.hpp:824]: I added it but note that the docs are minimal because this is just a standard variant

**Comment by pavel-kirienko** [comment]
Well, the pull request is finished, except that the CI is still misbehaving. I have already split the tests into six translation units to manage the memory consumption by the compiler. I can split it further if needed but I am not yet confident that this is going to help. The new error from the CI is this:

>Received request to deprovision: The request was cancelled by the remote provider

**Comment by pavel-kirienko** [comment]
@thirtytwobits I would like to add an exception for the coverage requirement for variant.hpp because currently it is computed per template instantiation, and the test suite is built to test a small fraction of branches per instantiation, which leads to a very low coverage estimate.

image

**Review by thirtytwobits — DISMISSED** [review]
Approved but could you add this example first: 

cetlvast/suites/docs/examples/example_09_variant.cpp

```
/// @file
/// Example of using cetl::variant.
///
/// @copyright
/// Copyright (C) OpenCyphal Development Team  <opencyphal.org>
/// Copyright Amazon.com Inc. or its affiliates.
/// SPDX-License-Identifier: MIT
///
#include "cetl/pf17/variant.hpp"
#include <string>
#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>

#include <gtest/gtest.h>

TEST(example_09_variant, basic_usage)
{
//! [example_09_basic_usage]
/// This example is taken directly from the [cppreference.com](https://en.cppreference.com/w/cpp/utility/variant)
/// documentation.
    cetl::pf17::variant<int, float> v, w;
    v = 42; // v contains int
    int i = cetl::pf17::get<int>(v);
    assert(42 == i); // succeeds
    w = cetl::pf17::get<int>(v);
    w = cetl::pf17::get<0>(v); // same effect as the previous line
    w = v; // same effect as the previous line

//  cetl::pf17::get<double>(v); // error: no double in [int, float]
//  cetl::pf17::get<3>(v);      // error: valid index values are 0 and 1

    float* result = cetl::pf17::get_if<float>(&w); // w contains int, not float: will return null
    std::cout << "get_if<float>(&w) => " << result << '\n';

    using namespace std::literals;

    cetl::pf17::variant<std::string> x("abc");
    // converting constructors work when unambiguous
    x = "def"; // converting assignment also works when unambiguous

    cetl::pf17::variant<std::string, void const*> y("abc");
    // casts to void const * when passed a char const *
    assert(cetl::pf17::holds_alternative<void const*>(y)); // succeeds
    y = "xyz"s;
    assert(cetl::pf17::holds_alternative<std::string>(y)); // succeeds
//! [example_09_basic_usage]
}
```

**Pull request merged**